lib/gauge_polyakov_loop.cu

```diff
@@ -0,0 +1,276 @@
+      // multiplies for each loop plus traces
+      return mat_mul_flops * u.Volume() / u.Geometry() + 2 * Nc * u.Volume() / u.X()[trace_direction];
+    }
+
+    long long bytes() const {
+      // links * one LatticeColorMatrix
+      return u.Bytes() / u.Geometry();
+    }
+  };
+
+  /**
+   * @brief Instantiation helper for the routine that inserts 3-d fields into a single timeslice
+   * of a 4-d field
+   *
+   * @param[out] u 4-d gauge field
+   * @param[in] ploop Input 3-d gauge field to be inserted
+   * @param[in] timeslice Timeslice of the 4-d field in which to insert the 3-d gauge field
+   */
+  void gaugeInsertTimeslice(GaugeField &u, GaugeField &s, int timeslice) {
+    instantiate<GaugeInsertTimeslice>(u, s, timeslice);
+  }
+
+  void gaugePolyakovLoop(double ploop[2], const GaugeField& u, int dir, TimeProfile &profile) {
+
+    if (dir != 3) errorQuda("Unsupported direction %d", dir);
+
+    // output array
+    array<double, 2> loop;
+
+    std::unique_ptr<GaugeField> condensed_field;
+
+    // If the dir dimension isn't partitioned, we can just do a quick compute + reduce,
+    // otherwise we need to perform a two-step calculation: one per-GPU, then gather
+    // and do the final product + reduce.
+    if (commDimPartitioned(dir)) {
+
+      // Form a staging gauge field where each "t" slice corresponds to one rank in the "dir"
+      // direction. Odd dimensions are carefully legal in the `t` dimension in this context
+      profile.TPSTART(QUDA_PROFILE_INIT);
+      GaugeFieldParam gParam(u);
+      lat_dim_t x;
+      for (int d = 0; d < 3; d++) x[d] = u.X()[d];
+      x[3] = 1;
+      gParam.x = x;
+      gParam.create = QUDA_NULL_FIELD_CREATE;
+      gParam.ghostExchange = QUDA_GHOST_EXCHANGE_NO;
+      gParam.location = QUDA_CUDA_FIELD_LOCATION;
+      gParam.geometry = QUDA_SCALAR_GEOMETRY;
+
+      // We promote the precision to double to maintain sufficient reproducibility
+      // as a function of the number of ranks in the `t` dimension
+      gParam.setPrecision(QUDA_DOUBLE_PRECISION);
+
+      std::unique_ptr<GaugeField> product_field = std::make_unique<cudaGaugeField>(gParam);
+      GaugeField& product_field_ref = reinterpret_cast<GaugeField&>(*product_field.get());
+
+      // Create the field we reduce into
+      x[3] = comm_dim(3);
+      gParam.x = x;
+      gParam.create = QUDA_NULL_FIELD_CREATE;
+      condensed_field = std::make_unique<cudaGaugeField>(gParam);
+      GaugeField& condensed_field_ref = reinterpret_cast<GaugeField&>(*condensed_field.get());
+      profile.TPSTOP(QUDA_PROFILE_INIT);
+
+      profile.TPSTART(QUDA_PROFILE_COMPUTE);
+
+      // Compute the rank-local product of loops in the temporal direction...
+      instantiate<GaugePolyakovLoopProduct, ReconstructNo12>(u, product_field_ref);
+      // ...and inject it into my own accumulation field
+      gaugeInsertTimeslice(condensed_field_ref, product_field_ref, comm_coord(3));
+      profile.TPSTOP(QUDA_PROFILE_COMPUTE);
```

Comment:
The perhaps unusual aspect of the below is that it's in a .cu file.  Possible to removes this comms handling stuff into a new cpp file?

Reply:
To be more coherent here: I am wondering if this communications pattern should be abstracted into a separate function in its own cpp file as opposed to being part of the Polyakov loop driver.

Reply:
Discussed online, we can revisit this if the need for a ring buffer pops up elsewhere, but there's no real need now.